### ISSUE TITLE
BUG: np.cross() warns about arrays of vectors when used with two simple 2-d vectors

### ISSUE THREAD

**Issue by megies** [numpy/numpy#26620]
### Describe the issue:

`np.cross()` warns about arrays of vectors when used with simply two 2-d vectors (this use case is even in the ["Examples" section of `np.cross()` docstring](https://github.com/numpy/numpy/blob/42e9aa2c8ce6791cf175bd1eeec46d533c7c41d1/numpy/_core/numeric.py#L1591-L1596)). I believe, that check whether a warning should be shown is missing a check for number of dimensions of the input.

https://github.com/numpy/numpy/blob/42e9aa2c8ce6791cf175bd1eeec46d533c7c41d1/numpy/_core/numeric.py#L1645-L1653

I feel like there should be a change, either
- change the warning message to something like `'2-dimensional vectors are deprecated. Use 3-dimensional vectors instead.'` without the `'Arrays of..'` parts, or..
- the check whether the warning should be shown or not should also check `if a.ndim > 1 or b.ndim > 1`

Tbh, I'm a bit confused what that warning is implying as recommended usage anyway. Is the user supposed to add an all-zero third dimension for 2-d vectors and then calculate cross product in 3-d and then extract the valid part from it? That seems kinda odd-looking, I guess if that really is the case it looks like this prompts to look at #13718 again potentially?

So instead of..

```python
x = [[1, 2], [1, 2]]
y = [[4, 5], [4, 5]]
result = np.cross(x, y)
```

user should do something like..?

```python
x = [[1, 2, 0], [1, 2, 0]]
y = [[4, 5, 0], [4, 5, 0]]
result = np.cross(x, y)[:, -1]
```

### Reproduce the code example:

```python
import numpy as np

x = [1, 2]
y = [4, 5]
# the next line shows a warning, even though it's even from the docstring examples section
np.cross(x, y)
```

### Error message:

```shell
DeprecationWarning: Arrays of 2-dimensional vectors are deprecated. Use arrays of 3-dimensional vectors instead. (deprecated in NumPy 2.0)
```

### Python and NumPy Versions:

2.0.0rc2
3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:45:18) [GCC 12.3.0]

### Runtime Environment:

[{'numpy_version': '2.0.0rc2',
  'python': '3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:45:18) '
            '[GCC 12.3.0]',
  'uname': uname_result(system='Linux', node='mother', release='5.10.0-25-amd64', version='#1 SMP Debian 5.10.191-1 (2023-08-16)', machine='x86_64')},
 {'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2',
                                'AVX512F',
                                'AVX512CD',
                                'AVX512_SKX',
                                'AVX512_CLX',
                                'AVX512_CNL',
                                'AVX512_ICL'],
                      'not_found': ['AVX512_KNL', 'AVX512_KNM', 'AVX512_SPR']}},
 {'architecture': 'SkylakeX',
  'filepath': '/home/megies/miniconda3/envs/np2/lib/libopenblasp-r0.3.27.so',
  'internal_api': 'openblas',
  'num_threads': 8,
  'prefix': 'libopenblas',
  'threading_layer': 'pthreads',
  'user_api': 'blas',
  'version': '0.3.27'}]

### Context for the issue:

In [our](https://github.com/obspy/obspy/) testsuite, we also have a test runner with `pytest -W error` to make sure our code is not causing deprecation warnings in order to keep our code base work with upcoming releases of dependencies. And that's where this deprecation warning came up.

**Comment by seberg** [comment]
Ref https://github.com/numpy/numpy/pull/24818, does anyone know what the argument/reasoning for this was?
* Avoid variable number of result dimensions?
* Deprecated `cross([0, 1], [1, 2, 3])` from working?

**Comment by seberg** [comment]
OK, since I missed it there. The issue is the first thing and the reasoning is in https://github.com/numpy/numpy/issues/13718 which suggested a `cross2d`, though.

**Comment by seberg** [comment]
We had a bit of a discussing, we should definitely tell people to use this (I think it is right):
```
cross2d = arr1[..., 0] * arr2[..., 1] - arr1[..., 1] * arr2[..., 0]
```
And maybe even document it in the text.  The original issue had `cross2d`, and if a lot of people use the 2d version, then it might be nice to have it as `np.linalg.cross2d`.  If that was clear now, I might prefer to not do the deprecation as quickly, because you cannot use `cross2d` until it exists (and is established).

**Comment by rgommers** [comment]
> The original issue had `cross2d`, and if a lot of people use the 2d version, then it might be nice to have it as `np.linalg.cross2d`. If that was clear now, I might prefer to not do the deprecation as quickly, because you cannot use `cross2d` until it exists (and is established).

It looks like `linalg.cross2d` is moving ahead, although the need is still slightly on the thin side. I'm -1 on changing our mind on the deprecation at the very last moment, since that has downsides and since clearly this isn't used a lot - it took ~6 months after gh-25145 got merged and almost 3 months after `2.0.0b1` was out for someone to notice. 

Let's verify the one-liner above as a workaround (it seems right), and add that to the docstring as a compat note, and backport it to the 2.0.x branch. That should be easy and safe to do in the next few days.

**Comment by rgommers** [comment]
Once gh-26692 is backported to the 2.0.x branch, the milestone on this can be removed.

**Comment by Daniel63656** [comment]
Having the same issue. Why wouldn't cross() be abstract enough to handle 2d and 3d without the user having to differentiate

**Comment by leycec** [comment]
**lol.** So, [`numpy.linalg.cross2d()` never happened](https://github.com/numpy/numpy/pull/26640). Now `numpy.linalg.cross()` emits unhelpful deprecation warnings for the previously valid two-dimensional use case that fail to actually inform the end user of how to resolve the deprecation. If a use case requires two dimensions, telling end users to just use three dimensions instead isn't really a useful recommendation. In short, this is basically useless:

```python
DeprecationWarning: Arrays of 2-dimensional vectors are deprecated.
Use arrays of 3-dimensional vectors instead. (deprecated in NumPy 2.0)
```

That's a hard UX fail. Simply hand-waving that away as "fine for certain definitions of fine" because [users failed to test against NumPy pre-releases like `2.0.0b1`](https://github.com/numpy/numpy/issues/26620#issuecomment-2163209159) isn't a reasonable justification, either. Of *course* users failed to test against NumPy pre-releases. They're pre-releases! Without compelling reasons to install third-party pre-releases, users (*...waitforit*) do not install third-party pre-releases. Why would they? It's needless work for no real-world gain and significant pain.

## tl;dr

In short, the existing `DeprecationWarning`:

* **(A)** Is less than useless. 2D code doesn't magically become 3D code just because you tell it to become 3D code. Yes, that's exactly what the existing `DeprecationWarning` is telling everybody. `</facepalm>`
* **(B)** Should be entirely rewritten to reflect [the trivial one-liner identified above](https://github.com/numpy/numpy/issues/26620#issuecomment-2150748569) by @seberg. Something resembling the following would actually be useful: e.g.,

```python
DeprecationWarning: np.cross() no longer supports 2-dimensional vectors.
Replace all np.cross() calls over 2-dimensional vectors with one-liners:
    cross2d = arr1[..., 0] * arr2[..., 1] - arr1[..., 1] * arr2[..., 0]
```

I shake my head, fam. I shake it hard. :face_exhaling:

**Comment by burnpanck** [comment]
Pity that `np.linalg.cross2d` never happened. While the oneliner is trivial to write when the inputs are arrays, if the inputs are expressions, the one-liner becomes a three-liner! Its of course doable, but having a true one-liner for the "2d cross product" around is a great nice-to-have for any 2D geometry task.

**Comment by cslotboom** [comment]
This has also affected me. Converting all my 2D vectors to 3D vectors with an empty row just to cross multiply feels like a pretty clunky work around, I'd strongly prefer not to do that if it's in the cards.

**Comment by assertivist** [comment]
Hello, I would like to bump this issue. From a library usage perspective, I have code that previously (and continues) to work fine, but now fills up the `stderr` of my users with this unhelpful message that doesn't apply to the code it points to (no arrays). 

I also would strongly prefer not to convert my vector _expressions_ to 3D, it would make my code _especially_ confusing, because they are converted to 2D _specifically, just before_ the cross-multiply.


If I'm going to patch this I'm going to remove the call to `cross()` altogether, adding like twelve lines of code so you all can save two to four lines. 

Maybe I am missing something, but does the one-liner's "triviality" not imply that it should simply be included in the library to handle the case of 2 dimensional vectors being passed to the cross-multiply function? `np.math.factorial` can be replaced by `lambda n:1 if n<2 else n*f(n-1)` so are we getting rid of that one too?

**Comment by mattip** [comment]
Reviewing all this:

The original motivation for disallowing 2d `cross` was to be compatible with the [Array API definition](https://data-apis.org/array-api/latest/extensions/generated/array_api.linalg.cross.html). If I understand correctly this was discussed and arrived on in [an Array API issue](https://github.com/data-apis/array-api/issues/415), based partially on #13718 and the [comment](https://github.com/numpy/numpy/issues/13718#issuecomment-960453783) about the different signature of the two functions:
> `cross2` has signature `(2),(2)->()`, and `cross3` has signature `(3),(3)->(3)`. 

The PR #26640 to add a separate `linalg.cross2d` seems to have fizzled out before being closed. Maybe someone who has an interest would be willing to revive that PR in some form, maybe with a new implementation?
